server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -78,21 +78,24 @@ public static FieldFetcher create(QueryShardContext context,
                 Regex.simpleMatchToAutomaton(unmappedFetchPattern.toArray(new String[unmappedFetchPattern.size()]))
             );
         }
-        return new FieldFetcher(fieldContexts, unmappedFetchAutomaton, mappedToExclude);
+        return new FieldFetcher(fieldContexts, unmappedFetchAutomaton, mappedToExclude, includeUnmapped);
     }
 
     private final List<FieldContext> fieldContexts;
     private final CharacterRunAutomaton unmappedFetchAutomaton;
     private final Set<String> mappedToExclude;
+    private final boolean includeUnmapped;
 
     private FieldFetcher(
         List<FieldContext> fieldContexts,
         CharacterRunAutomaton unmappedFetchAutomaton,
-        Set<String> mappedToExclude
+        Set<String> mappedToExclude,
```

Comment:
I just noticed -- maybe we could switch `fieldContexts` to a `LinkedHashMap`? Then we could consult its key set and avoid creating this new set `mappedToExclude`. I could tackle this in a follow-up though, to keep this PR small.

Reply:
Interesting idea, I took a quick look and wasn't sure if we have to consider edge cases where the current list might contain different elements with the same field key though. I agree we should probably do this as a follow up in the interest of getting the basics in this PR in first.